File: react_native_app/src/services/ApiService.js
/**
 * API Service for WMS Mobile App
 * Handles all HTTP communication with Flask backend
 * Supports MySQL database integration as per user preference
 */

import axios from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { API_CONFIG } from '../config/config';

class ApiServiceClass {
  constructor() {
    // Create axios instance with base configuration
    this.api = axios.create({
      baseURL: API_CONFIG.BASE_URL,
      timeout: API_CONFIG.TIMEOUT,
      headers: {
        'Content-Type': 'application/json',
      },
    });

    // Add request interceptor to include auth token
    this.api.interceptors.request.use(
      async (config) => {
        const token = await AsyncStorage.getItem('userToken');
        if (token) {
          config.headers.Authorization = `Bearer ${token}`;
        }
        return config;
      },
      (error) => {
        return Promise.reject(error);
      }
    );

    // Add response interceptor for error handling
    this.api.interceptors.response.use(
      (response) => response,
      async (error) => {
        if (error.response?.status === 401) {
          // Token expired, logout user
          await AsyncStorage.removeItem('userToken');
          await AsyncStorage.removeItem('userData');
          // Redirect to login screen
        }
        return Promise.reject(error);
      }
    );
  }

  // Authentication endpoints
  async login(username, password) {
    try {
      const response = await this.api.post('/auth/login', {
        username,
        password,
      });
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async logout() {
    try {
      await this.api.post('/auth/logout');
    } catch (error) {
      console.log('Logout error:', error);
    }
  }

  async healthCheck() {
    try {
      const response = await this.api.get('/api/health');
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  // GRPO Module API
  async getGRPODocuments() {
    try {
      const response = await this.api.get('/api/grpo_documents');
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async getGRPODocument(id) {
    try {
      const response = await this.api.get(`/api/grpo_documents/${id}`);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async createGRPODocument(grpoData) {
    try {
      const response = await this.api.post('/api/grpo_documents', grpoData);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async updateGRPODocument(id, grpoData) {
    try {
      const response = await this.api.put(`/api/grpo_documents/${id}`, grpoData);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async submitGRPOForQC(id) {
    try {
      const response = await this.api.post(`/api/grpo_documents/${id}/submit`);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async approveGRPO(id, notes = '') {
    try {
      const response = await this.api.post(`/api/grpo_documents/${id}/approve`, {
        qc_notes: notes,
      });
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async rejectGRPO(id, notes = '') {
    try {
      const response = await this.api.post(`/api/grpo_documents/${id}/reject`, {
        qc_notes: notes,
      });
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  // Inventory Transfer Module API
  async getInventoryTransfers() {
    try {
      const response = await this.api.get('/api/inventory_transfers');
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async getInventoryTransfer(id) {
    try {
      const response = await this.api.get(`/api/inventory_transfers/${id}`);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async createInventoryTransfer(transferData) {
    try {
      const response = await this.api.post('/api/inventory_transfers', transferData);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async updateInventoryTransfer(id, transferData) {
    try {
      const response = await this.api.put(`/api/inventory_transfers/${id}`, transferData);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async submitTransferForQC(id) {
    try {
      const response = await this.api.post(`/api/inventory_transfers/${id}/submit`);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async approveTransfer(id, notes = '') {
    try {
      const response = await this.api.post(`/api/inventory_transfers/${id}/qc_approve`, {
        qc_notes: notes,
      });
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async rejectTransfer(id, notes = '') {
    try {
      const response = await this.api.post(`/api/inventory_transfers/${id}/qc_reject`, {
        qc_notes: notes,
      });
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async reopenTransfer(id) {
    try {
      const response = await this.api.post(`/api/inventory_transfers/${id}/reopen`);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  // Pick List Module API
  async getPickLists() {
    try {
      const response = await this.api.get('/api/pick_lists');
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async getPickList(id) {
    try {
      const response = await this.api.get(`/api/pick_lists/${id}`);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async createPickList(pickListData) {
    try {
      const response = await this.api.post('/api/pick_lists', pickListData);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async updatePickList(id, pickListData) {
    try {
      const response = await this.api.put(`/api/pick_lists/${id}`, pickListData);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  // Barcode and Validation API
  async validateBarcode(barcode) {
    try {
      const response = await this.api.post('/api/validate_barcode', {
        barcode,
      });
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async validateTransferRequest(requestNumber) {
    try {
      const response = await this.api.post('/api/validate_transfer_request', {
        request_number: requestNumber,
      });
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  async validatePurchaseOrder(poNumber) {
    try {
      const response = await this.api.post('/api/validate_purchase_order', {
        po_number: poNumber,
      });
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  // Error handling
  handleError(error) {
    if (error.response) {
      // Server responded with error status
      const message = error.response.data?.message || error.response.data?.error || 'Server error occurred';
      return new Error(message);
    } else if (error.request) {
      // Network error
      return new Error('Network error. Please check your connection.');
    } else {
      // Other error
      return new Error(error.message || 'An unexpected error occurred');
    }
  }
}

export const ApiService = new ApiServiceClass();

File: mobile_app/lib/services/api_service.dart
import 'package:dio/dio.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/user.dart';
import '../models/inventory_transfer.dart';
import '../models/grpo_document.dart';
import '../utils/constants.dart';

class ApiService {
  late final Dio _dio;
  String? _authToken;

  ApiService() {
    _dio = Dio(BaseOptions(
      baseUrl: Constants.apiBaseUrl,
      connectTimeout: const Duration(seconds: 15),
      receiveTimeout: const Duration(seconds: 15),
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
    ));

    // Add interceptors
    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) async {
          // Add auth token to requests
          if (_authToken != null) {
            options.headers['Authorization'] = 'Bearer $_authToken';
          }
          
          print('🌐 API Request: ${options.method} ${options.path}');
          handler.next(options);
        },
        onResponse: (response, handler) {
          print('✅ API Response: ${response.statusCode} ${response.requestOptions.path}');
          handler.next(response);
        },
        onError: (error, handler) {
          print('❌ API Error: ${error.response?.statusCode} ${error.requestOptions.path}');
          print('Error message: ${error.message}');
          handler.next(error);
        },
      ),
    );
  }

  // Authentication
  Future<Map<String, dynamic>> login(String username, String password) async {
    try {
      final response = await _dio.post('/auth/login', data: {
        'username': username,
        'password': password,
      });

      if (response.statusCode == 200) {
        final data = response.data;
        _authToken = data['token'];
        
        // Save token to shared preferences
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString('auth_token', _authToken!);
        await prefs.setString('user_data', data['user'].toString());
        
        return data;
      } else {
        throw Exception('Login failed');
      }
    } catch (e) {
      throw Exception('Login error: $e');
    }
  }

  Future<void> logout() async {
    try {
      await _dio.post('/auth/logout');
    } catch (e) {
      print('Logout error: $e');
    } finally {
      _authToken = null;
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove('auth_token');
      await prefs.remove('user_data');
    }
  }

  Future<bool> loadStoredToken() async {
    final prefs = await SharedPreferences.getInstance();
    _authToken = prefs.getString('auth_token');
    return _authToken != null;
  }

  // Inventory Transfers
  Future<List<InventoryTransfer>> getInventoryTransfers() async {
    try {
      final response = await _dio.get('/api/inventory_transfers');
      
      if (response.statusCode == 200) {
        final List<dynamic> data = response.data['transfers'] ?? response.data;
        return data.map((json) => InventoryTransfer.fromJson(json)).toList();
      } else {
        throw Exception('Failed to fetch inventory transfers');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  Future<InventoryTransfer> getInventoryTransfer(int id) async {
    try {
      final response = await _dio.get('/api/inventory_transfers/$id');
      
      if (response.statusCode == 200) {
        return InventoryTransfer.fromJson(response.data);
      } else {
        throw Exception('Failed to fetch inventory transfer');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  Future<InventoryTransfer> createInventoryTransfer(Map<String, dynamic> data) async {
    try {
      final response = await _dio.post('/api/inventory_transfers', data: data);
      
      if (response.statusCode == 201) {
        return InventoryTransfer.fromJson(response.data);
      } else {
        throw Exception('Failed to create inventory transfer');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  Future<InventoryTransfer> updateInventoryTransfer(int id, Map<String, dynamic> data) async {
    try {
      final response = await _dio.put('/api/inventory_transfers/$id', data: data);
      
      if (response.statusCode == 200) {
        return InventoryTransfer.fromJson(response.data);
      } else {
        throw Exception('Failed to update inventory transfer');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  Future<Map<String, dynamic>> submitTransferForQC(int id) async {
    try {
      final response = await _dio.post('/api/inventory_transfers/$id/submit');
      
      if (response.statusCode == 200) {
        return response.data;
      } else {
        throw Exception('Failed to submit transfer for QC');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  Future<Map<String, dynamic>> reopenTransfer(int id) async {
    try {
      final response = await _dio.post('/api/inventory_transfers/$id/reopen');
      
      if (response.statusCode == 200) {
        return response.data;
      } else {
        throw Exception('Failed to reopen transfer');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  // GRPO Documents
  Future<List<GRPODocument>> getGRPODocuments() async {
    try {
      final response = await _dio.get('/api/grpo_documents');
      
      if (response.statusCode == 200) {
        final List<dynamic> data = response.data['grpos'] ?? response.data;
        return data.map((json) => GRPODocument.fromJson(json)).toList();
      } else {
        throw Exception('Failed to fetch GRPO documents');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  Future<GRPODocument> getGRPODocument(int id) async {
    try {
      final response = await _dio.get('/api/grpo_documents/$id');
      
      if (response.statusCode == 200) {
        return GRPODocument.fromJson(response.data);
      } else {
        throw Exception('Failed to fetch GRPO document');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  Future<GRPODocument> createGRPODocument(Map<String, dynamic> data) async {
    try {
      final response = await _dio.post('/api/grpo_documents', data: data);
      
      if (response.statusCode == 201) {
        return GRPODocument.fromJson(response.data);
      } else {
        throw Exception('Failed to create GRPO document');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  // Barcode Validation
  Future<Map<String, dynamic>> validateBarcode(String barcode) async {
    try {
      final response = await _dio.post('/api/validate_barcode', data: {
        'barcode': barcode,
      });
      
      if (response.statusCode == 200) {
        return response.data;
      } else {
        throw Exception('Failed to validate barcode');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  // Transfer Request Validation  
  Future<Map<String, dynamic>> validateTransferRequest(String requestNumber) async {
    try {
      final response = await _dio.post('/api/validate_transfer_request', data: {
        'request_number': requestNumber,
      });
      
      if (response.statusCode == 200) {
        return response.data;
      } else {
        throw Exception('Failed to validate transfer request');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  // QC Operations
  Future<Map<String, dynamic>> qcApproveTransfer(int id, String? notes) async {
    try {
      final response = await _dio.post('/api/inventory_transfers/$id/qc_approve', data: {
        'qc_notes': notes ?? '',
      });
      
      if (response.statusCode == 200) {
        return response.data;
      } else {
        throw Exception('Failed to approve transfer');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  Future<Map<String, dynamic>> qcRejectTransfer(int id, String reason) async {
    try {
      final response = await _dio.post('/api/inventory_transfers/$id/qc_reject', data: {
        'qc_notes': reason,
      });
      
      if (response.statusCode == 200) {
        return response.data;
      } else {
        throw Exception('Failed to reject transfer');
      }
    } catch (e) {
      throw Exception('API error: $e');
    }
  }

  // Health Check
  Future<bool> checkConnection() async {
    try {
      final response = await _dio.get('/api/health');
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }
}